database/pgsql/layer.go

```diff
@@ -18,300 +18,349 @@ import (
+	for i, ns := range namespaces {
+		detectorID := detectorMap.byValue[ns.By]
+		namespaceID := rawNamespaceIDs[i].Int64
+		if !rawNamespaceIDs[i].Valid {
+			return database.ErrMissingEntities
+		}
 
-	sort.Strings(listers)
-	keys := make([]interface{}, len(listers)*2)
-	for i, d := range listers {
-		keys[i*2] = id
-		keys[i*2+1] = d
+		dbLayerNamespaces = append(dbLayerNamespaces, dbLayerNamespace{layerID, namespaceID, detectorID})
 	}
 
-	_, err := tx.Exec(queryPersistLayerListers(len(listers)), keys...)
+	return tx.persistLayerNamespaces(dbLayerNamespaces)
+}
+
+func (tx *pgSession) persistAllLayerFeatures(layerID int64, features []database.LayerFeature) error {
+	detectorMap, err := tx.findAllDetectors()
 	if err != nil {
-		return handleError("queryPersistLayerDetectors", err)
+		return err
 	}
-	return nil
-}
 
-func (tx *pgSession) persistLayerFeatures(id int64, features []database.Feature) error {
-	if len(features) == 0 {
-		return nil
+	rawFeatures := make([]database.Feature, 0, len(features))
+	for _, f := range features {
+		rawFeatures = append(rawFeatures, f.Feature)
 	}
 
-	fIDs, err := tx.findFeatureIDs(features)
+	featureIDs, err := tx.findFeatureIDs(rawFeatures)
 	if err != nil {
 		return err
 	}
 
-	ids := make([]int, len(fIDs))
-	for i, fID := range fIDs {
-		if !fID.Valid {
-			return errNamespaceNotFound
+	dbFeatures := make([]dbLayerFeature, 0, len(features))
+	for i, f := range features {
+		detectorID := detectorMap.byValue[f.By]
+		featureID := featureIDs[i].Int64
+		if !featureIDs[i].Valid {
+			return database.ErrMissingEntities
 		}
-		ids[i] = int(fID.Int64)
-	}
 
-	sort.IntSlice(ids).Sort()
-	keys := make([]interface{}, len(features)*2)
-	for i, fID := range ids {
-		keys[i*2] = id
-		keys[i*2+1] = fID
+		dbFeatures = append(dbFeatures, dbLayerFeature{layerID, featureID, detectorID})
 	}
 
-	_, err = tx.Exec(queryPersistLayerFeature(len(features)), keys...)
-	if err != nil {
-		return handleError("queryPersistLayerFeature", err)
+	if err := tx.persistLayerFeatures(dbFeatures); err != nil {
+		return err
 	}
+
 	return nil
 }
 
-func (tx *pgSession) persistLayerNamespace(id int64, namespaces []database.Namespace) error {
-	if len(namespaces) == 0 {
+func (tx *pgSession) persistLayerFeatures(features []dbLayerFeature) error {
+	if len(features) == 0 {
 		return nil
 	}
 
-	nsIDs, err := tx.findNamespaceIDs(namespaces)
-	if err != nil {
-		return err
-	}
-
-	// for every bulk persist operation, the input data should be sorted.
-	ids := make([]int, len(nsIDs))
-	for i, nsID := range nsIDs {
-		if !nsID.Valid {
-			panic(errNamespaceNotFound)
-		}
-		ids[i] = int(nsID.Int64)
-	}
+	sort.Slice(features, func(i, j int) bool {
+		return features[i].featureID < features[j].featureID
+	})
 
-	sort.IntSlice(ids).Sort()
-
-	keys := make([]interface{}, len(namespaces)*2)
-	for i, nsID := range ids {
-		keys[i*2] = id
-		keys[i*2+1] = nsID
+	keys := make([]interface{}, len(features)*3)
```

Comment:
TODO: abstract this, it's hard for me to use!

Reply:
it's like reading C 😎